### PR TITLE
DB Constraints (currently on UNIQUE constraint)

### PR DESCRIPTION
**Pull request by YingjingLu**
1. Documentation file: P3DesignDoc.md in the root of the repo
2. The current pull request focuses on the UNIQUE Constraint. We added creating UNIQUE constraint during the table creation process in the ddl_executor, which calls calalog_accessor's API to create UNIQUE constraint and insert it in a book-keeping table. There is also a GetConstraint() method to retrieve constraints for a table which is in database_catalog
3. Files changed to be reviewed:
`src/include/execution/sql/ddl_executors.h
src/execution/sql/ddl_executors.cpp
src/include/catalog/catalog_accessor.h
src/include/catalog/database_catalog.h 
src/catalog/database_catalog.cpp
src/include/catalog/postgres/pg_constraint.h
src/catalog/catalog_accessor.cpp
P3DesignDoc.md (Design Doc)`

### CODE REVIEW

File: P3DesignDoc.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
+# Database Constraints
```
Comment (tianhanhu) [inline P3DesignDoc.md:1]: Test comment

File: src/catalog/database_catalog.cpp [hunk 2]
```diff
@@ -953,11 +964,154 @@ const Schema &DatabaseCatalog::GetSchema(const common::ManagedPointer<transactio
...
+  // Write the namespace_oid into the PR
+  const auto con_namespace_oid_offset = pg_constraints_all_cols_prm_[postgres::CONNAMESPACE_COL_OID];
+  auto *const con_namespace_oid_ptr = constraints_insert_pr->AccessForceNotNull(con_namespace_oid_offset);
+  *(reinterpret_cast<namespace_oid_t *>(con_namespace_oid_ptr)) = ns_oid;
+
+  // Write constraint_type
+  const auto con_type_oid_offset = pg_constraints_all_cols_prm_[postgres::CONTYPE_COL_OID];
+  auto *const con_type_oid_ptr = constraints_insert_pr->AccessForceNotNull(con_type_oid_offset);
+  if (schema.is_unique_) {
+    *(reinterpret_cast<char *>(con_type_oid_ptr)) = 'u';
```
Comment (tianhanhu) [inline src/catalog/database_catalog.cpp:1003]: Might need some comments here reminding people what 'u', 'p' stand for.

File: src/catalog/database_catalog.cpp [hunk 3]
```diff
@@ -953,11 +964,154 @@ const Schema &DatabaseCatalog::GetSchema(const common::ManagedPointer<transactio
...
+  }
+
+  *(reinterpret_cast<bool *>(constraints_insert_pr->AccessForceNotNull(
+      pg_constraints_all_cols_prm_[postgres::CONDEFERRABLE_COL_OID]))) = false;
+  *(reinterpret_cast<bool *>(constraints_insert_pr->AccessForceNotNull(
+      pg_constraints_all_cols_prm_[postgres::CONDEFERRED_COL_OID]))) = false;
+  *(reinterpret_cast<bool *>(constraints_insert_pr->AccessForceNotNull(
+      pg_constraints_all_cols_prm_[postgres::CONVALIDATED_COL_OID]))) = true;
+
+//  const auto con_relid_oid_offset = pg_constraints_all_cols_prm_[postgres::CONRELID_COL_OID];
```
Comment (tianhanhu) [inline src/catalog/database_catalog.cpp:1015]: Not sure if this code is going to be reused in the future? Might consider removing it if it is dead code?

File: src/catalog/database_catalog.cpp [hunk 4]
```diff
@@ -953,11 +964,154 @@ const Schema &DatabaseCatalog::GetSchema(const common::ManagedPointer<transactio
...
+//  index_pr = constraints_name_index_init.InitializeRow(index_buffer);
+//  *(reinterpret_cast<table_oid_t *>(index_pr->AccessForceNotNull(0))) = table_oid;
+//  if (!indexes_table_index_->Insert(txn, *index_pr, indexes_tuple_slot)) {
+//    // There was duplicate value. Free the buffer and
+//    // return INVALID_TABLE_OID to indicate the database was not created.
+//    delete[] index_buffer;
+//    return false;
+//  }
+
+//  // Free the buffer, we are finally done
```
Comment (tianhanhu) [inline src/catalog/database_catalog.cpp:1057]: Same here with the potential dead code.

File: src/include/catalog/postgres/pg_constraint.h [hunk 5]
```diff
@@ -51,4 +51,172 @@ enum class ConstraintType : char {
...
+  }
+
+  // create a new FK constraint
+  PG_Constraint *GetNewFKConstraint(namespace_oid_t namespace_id, ConstraintType con_type, table_oid_t con_table_id,
+                                    col_oid_t col_id, table_oid_t ref_table, col_oid_t ref_col, bool update_cascade,
+                                    bool delete_cascade) {
+    constraint_oid_t tmp_oid;
+    PG_Constraint *con;
+    latch_.Lock();
+    if (!unused_con_oid_.empty()) {
```
Comment (tianhanhu) [inline src/include/catalog/postgres/pg_constraint.h:186]: This part of code seems duplicated with the part in GetNewConstraint.
Could consider putting it in a helper function for finding next con_oid.

**Comment by zianke** [comment]
Overall your implementation looks very good. But I agree with Tianhan that unit tests should be added to ensure the correctness. They can also give reviewers more examples of how your code is expected to work. 
By the way, just a few small suggestions on making a pull request. It might be better to keep only the necessary code changes in a pull request, considering it will potentially be merged into master. I would suggest not updating files like `.gitignore` (you can edit `.git/info/exclude` instead) and put the design document into the repository's [Wiki](https://github.com/YingjingLu/721-P3-Constraints/wiki).

File: src/catalog/database_catalog.cpp [hunk 6]
```diff
@@ -953,11 +964,153 @@ const Schema &DatabaseCatalog::GetSchema(const common::ManagedPointer<transactio
   return *reinterpret_cast<Schema *>(ptr_pair.first);
 }
 
+constraint_oid_t DatabaseCatalog::CreateConstraints(const common::ManagedPointer<transaction::TransactionContext> txn,
+                                         namespace_oid_t ns, const std::string &name, table_oid_t table,
+                                         const IndexSchema &schema) {
+  if (!TryLock(txn)) return INVALID_CONSTRAINT_OID;
+  const constraint_oid_t constraint_oid_t = static_cast<::terrier::catalog::constraint_oid_t>(next_oid_++);
```
[the remark this answers is not in the document]
Comment (apavlo) [reply src/catalog/database_catalog.cpp:971]: @adrianchang is correct. This is like naming an integer variable `int`.

File: src/catalog/database_catalog.cpp (hunk 2 above)
Comment (apavlo) [reply src/catalog/database_catalog.cpp]: @tianhanhu is correct! These should be using `ConstraintType`

**Comment by apavlo** [comment]
Closing for now to reduce Jenkins contention.

**Pull request closed without merging**